Dnnl training

georgen117 opened the pull request:
**Description**: Describe your changes.
This adds training ops ReluGrad, ConvGrad, and MaxPoolGrad to enable training using dnnl (a.k.a OneDNN).

**Motivation and Context**
Expand the ops covered by the dnnl_provider by enabling training.

jywu-msft commented:
/azp run Linux DNNL CI Pipeline

jywu-msft commented:
/azp run orttraining-linux-ci-pipeline

jywu-msft commented:
/azp run orttraining-linux-gpu-ci-pipeline

jywu-msft commented:
/azp run Linux CPU CI Pipeline

jywu-msft commented:
@mrry , @SherlockNoMad, would you guys be able to take a quick look?

snnn commented:
@georgen117 Could you please fix the conflicts?

georgen117 commented:
yes, I can fix the conflicts.

mrry commented on onnxruntime/test/providers/provider_test_utils.cc:
```diff
@@ -769,7 +769,35 @@ void OpTester::Run(
         provider_types += entry->Type() + ":";
-        ASSERT_STATUS_OK(session_object.RegisterExecutionProvider(std::move(entry)));
+
+        std::unique_ptr<IExecutionProvider> execution_provider;
+        if (entry->Type() == onnxruntime::kCpuExecutionProvider)
+          execution_provider = DefaultCpuExecutionProvider();
+        else if (entry->Type() == onnxruntime::kCudaExecutionProvider)
+          execution_provider = DefaultCudaExecutionProvider();
+        else if (entry->Type() == onnxruntime::kDnnlExecutionProvider)
+          execution_provider = DefaultDnnlExecutionProvider();
+        else if (entry->Type() == onnxruntime::kOpenVINOExecutionProvider)
+          execution_provider = DefaultOpenVINOExecutionProvider();
+        else if (entry->Type() == onnxruntime::kNupharExecutionProvider)
+          execution_provider = DefaultNupharExecutionProvider();
+        else if (entry->Type() == onnxruntime::kTensorrtExecutionProvider)
+          execution_provider = DefaultTensorrtExecutionProvider();
+        else if (entry->Type() == onnxruntime::kNnapiExecutionProvider)
+          execution_provider = DefaultNnapiExecutionProvider();
+        else if (entry->Type() == onnxruntime::kRknpuExecutionProvider)
+          execution_provider = DefaultRknpuExecutionProvider();
+        else if (entry->Type() == onnxruntime::kAclExecutionProvider)
+          execution_provider = DefaultAclExecutionProvider();
+        else if (entry->Type() == onnxruntime::kArmNNExecutionProvider)
+          execution_provider = DefaultArmNNExecutionProvider();
+        else if (entry->Type() == onnxruntime::kRocmExecutionProvider)
+          execution_provider = DefaultRocmExecutionProvider();
+        // skip if execution provider is disabled
+        if (execution_provider == nullptr)
+          continue;
+
+        ASSERT_STATUS_OK(session_object.RegisterExecutionProvider(std::move(execution_provider)));
```
Why is this change necessary? It looks like it changes the meaning of the `execution_providers` argument, by switching the (potentially custom) passed-in object to a default instance of the same type.

georgen117 replied:
The gradient unit tests use the Jacobian matrix to validate the gradient function.  The tests end up calling the OpTester::Run function multiple times.

Before adding the dnnl execution provider to the gradient tests the code always executed the else branch when running the gradient tests. Avoiding this block of code.  The problem is that the `entry->Type()` is moved into the `session_object` the first time it is called.  The means the next time the Run function is called the memory found in `entry->Type()` has been moved resulting in an error.

This change creates a new execution provider based on the value found in entry->Type() and then moves the new execution provider in to the `session_object.RegisterExecutionProvider( ... )` leaving the `entry->Type()` unchanged so it can be referenced for the next time the `Run` function is called.

mrry replied:
Hmm, the problem here is that `entry->Type()` isn't sufficient to reconstruct all execution providers. For example, a `CudaExecutionProvider` might be constructed with a [`CudaExecutionProviderInfo`](https://github.com/microsoft/onnxruntime/blob/64709b1335380259b6b6e40cd02f5c3aeee7b38d/onnxruntime/core/providers/cuda/cuda_execution_provider_info.h) that is different from the default, and this change would silently drop those settings.

I think I understand the problem though... the documentation for the `execution_provider` argument isn't clear about the fact that the passed-in providers are consumed by the method (and moved into the session). Is it possible to fix this by recreating the dnnl EP each time this method is called?

Or does the problem arise because the gradient tests make  multiple calls to `OpTester::Run()` internally? :(. In that case, perhaps we should pass an optional vector of factory functions (`std::vector<std::function<std::unique_ptr<IExecutionProvider>>>*` ) that can recreate the EP(s) as necessary?

georgen117 replied:
> Is it possible to fix this by recreating the dnnl EP each time this method is called?
>
> Or does the problem arise because the gradient tests make multiple calls to OpTester::Run() internally?

None of the code written by me calls the `OpTester::Run() ` directly it is done indirectly somewhere in the test setup. Because of that, I don't think I could recreate the dnnl EP for each method call.

mrry replied:
OK, I've thought about this some more, and I think the root of the problem is that `GradientOpTester::Run()` (for some unknown reason) has an unused `execution_providers` argument, and attempting to use it causes problems because the execution providers end up being consumed multiple times. This part of the code looks like it needs to be refactored, but I wouldn't want that to derail this PR.

Given that, is it feasible to:
1. Revert the change to `OpTester::Run()` to avoid the breaking change, and
2. Localize the recreate-based-on-type logic to `GradientOpTester::Run()`?

If the tests don't pass with this suggestion, I'd prefer to disable the failing tests with a TODO/file an issue, rather than potentially break other callers of `OpTester::Run()`.

georgen117 replied:
I reverted the change and all of the tests completed without the segfault error that was the original reason for making the change. So I was able to remove the lines in question and all the tests pass.

Yay, we still have the test coverage without introducing the issue you were concerned with.

mrry commented on onnxruntime/test/testdata/onnx_backend_test_series_filters.jsonc:
```diff
@@ -68,6 +68,8 @@
         "^test_maxpool_2d_ceil",
         "^test_maxpool_2d_dilations",
         "^test_maxpool_2d_uint8",
+        "^test_maxpool_with_argmax_2d_precomputed_pads_cpu", // maxpool test will fail if built with --traning-enabled
+        "^test_maxpool_with_argmax_2d_precomputed_strides_cpu", // maxpool test will fail if built with --traning-enabled
```
Are these exclusions only needed for a DNNL build? If so, can we find some other way to disable them to avoid losing test coverage?

georgen117 replied:
These should only be excluded for the DNNL build these are listed under the `current_failing_tests_DNNL`. My understanding was that this was used if the `--use_dnnl` build flag was specified. I will double check.

georgen117 replied:
I just verified that when the code is built without the `--use-dnnl` flag tests are run.  So they are only disabled for dnnl.

mrry replied:
Apologies for the noise! I overlooked the context that these were added under the key `"current_failing_tests_DNNL"`.

georgen117 replied:
> Apologies for the noise! I overlooked the context that these were added under the key `"current_failing_tests_DNNL"`.

All good.  it is easy to miss since you only see the changed lines. I would rather have a more complete review with some noise.

mrry reviewed (changes requested):
Thank you for submitting this change! It's a large PR, and so I have a lot of comments :).

I've focused on the framework-level parts of the change, and haven't reviewed all of the kernels in detail. Hopefully someone who is more familiar with the DNNL EP can have a look at the kernels?

georgen117 commented:
> Thank you for submitting this change! It's a large PR, and so I have a lot of comments :).

@mrry I know this is a large PR. It is actually much larger than I typically feel comfortable with.  Feel free to leave as many comments as needed.

Thanks for the feedback.

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.cc:
```diff
@@ -129,11 +129,35 @@ void DNNLExecutionProvider::CreateOrUpdateDnnlNode(const Provider_Node* node,
   if (!fused) {
     ort_dnnl::DnnlNode dnnl_node;
     dnnl_node.name = node->OpType();
+// When running training mode the backward pass will need to access the forwardpass
```
I ran `clang-format` over just `dnnl_execution_provider.cc` it resulted in much more changes than expected. I think including all of the changes would result in an already large pull request being harder to review. I there a way to limit `clang-format` just lines there were modified?

georgen117 replied:
> Is there a way to limit `clang-format` just lines there were modified?

I figured out a way to limit the format changes to just the new lines.  Maybe in a future commit, I could do a bulk clang-format update to the dnnl provider code.  I don't want to make the code review harder for this pull request.

mrry replied:
Appreciate it, thanks! I'd be happy to approve a subsequent bulk update PR.

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.cc:
```diff
@@ -236,18 +260,25 @@ std::vector<std::unique_ptr<Provider_ComputeCapability>> DNNLExecutionProvider::
+#endif // !ENABLE_TRAINING
+      // TODO: Support this in training phase so that a valid entry would be added to the forward kernel map for this fusion. Without this, it would error out
+      // in the respective backward pass
+#ifndef ENABLE_TRAINING
```
Agree. Done.

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.cc:
```diff
@@ -389,8 +420,19 @@ void DNNLExecutionProvider::CreateMetaDef(const onnxruntime::Provider_GraphViewe
     auto itr = std::find(sub_var.outputs_as_input_other_node.begin(),
                          sub_var.outputs_as_input_other_node.end(), mklnode.output_name);
     if (itr == sub_var.outputs_as_input_other_node.end()) {
+#ifndef ENABLE_TRAINING
```
Moved the line out of #ifndef block.  Good catch.

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.h:
```diff
@@ -80,7 +86,21 @@ class DNNLExecutionProvider : public Provider_IExecutionProvider {
+  // Note if the a DnnlKernel already exists this will replace the existing kernel with the
+  // new kernel. This was done so the latest kernel is always placed in the map.
+  void SetForwardKernel(onnxruntime::NodeIndex key, std::shared_ptr<ort_dnnl::DnnlKernel> kernel) {
+    fwd_kernal_map_[key] = kernel;
```
Thanks.  Embarrassing typo.

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.h:
```diff
@@ -80,7 +86,21 @@ class DNNLExecutionProvider : public Provider_IExecutionProvider {
+    fwd_kernal_map_[key] = kernel;
+  }
 
+  // Fetch the kerenel using the NodeIndex
```
thanks fixed

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.h:
```diff
@@ -80,7 +86,21 @@ class DNNLExecutionProvider : public Provider_IExecutionProvider {
+  }
 
+  // Fetch the kerenel using the NodeIndex
+  std::shared_ptr<ort_dnnl::DnnlKernel> GetForwardKernal(onnxruntime::NodeIndex key) {
```
fixed

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.h:
```diff
@@ -94,6 +114,13 @@ class DNNLExecutionProvider : public Provider_IExecutionProvider {
 
+#ifdef ENABLE_TRAINING
+  // map used to hold and lookup forward DnnlKernels. This should only be needed in when
+  // running in training mode.The backward Kernels need access the forward kernals; typically
```
Fixed

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_conv.h:
```diff
@@ -258,6 +258,22 @@ class DnnlConv : public DnnlKernel {
 
     dnnl::memory::dims conv_zero_padding = {0, 0};
 
+#ifdef ENABLE_TRAINING
```
Yes. Much more readable.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_conv.h:
```diff
@@ -445,9 +462,9 @@ class DnnlConv : public DnnlKernel {
-    const T* filter_data = const_cast<T*>(ort.GetTensorData<T>(input_tensor));
 
     const int group_mkl = static_cast<int>(group_);
+    const T* filter_data = const_cast<T*>(ort.GetTensorData<T>(input_tensor));
```
Sorry about that. Reverted.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_func_kernel.cc:
```diff
@@ -186,6 +203,59 @@ class SubgraphPrimitive : public PrimitiveBase {
+        for (auto iter = dnnl_node.input_nodes.begin(); iter != dnnl_node.input_nodes.end(); ++iter) {
+          if (iter->op_type == "Relu") {
+            auto fwd_kernel = params.provider->GetForwardKernal(iter->index);
+            kernel->AddForwardDnnlKernel(std::dynamic_pointer_cast<DnnlRelu<T>>(fwd_kernel));
```
The number of input nodes is small but you are correct a `break` after finding the node would be more efficient.

georgen117 replied:
`break` added to the code.

mrry replied:
Thanks! Rather than efficiency, I prefer the `break` here to make it clear that `AddForwardDnnlKernel()` is only called at most once.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_func_kernel.cc:
```diff
@@ -259,7 +329,19 @@ template <typename T>
+    // each call. Since the SubgraphPrimitivePool stashes the nodes based on the thread_local memory it results in a new
+    // stash being created per-call from the training loop.  In theory the thread_local memory should be freed when the calling
+    // thread is destroyed but this was not being seen when actually running the code.  Instead of relying on the thread_local
+    // memory being freed we name a new SubgraphPrimive instead of using the SubgraphPrimitivePool when the code is built for
```
fixed

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h:
```diff
@@ -0,0 +1,258 @@
+    dnnl::engine engine_to_use;
+    std::unordered_map<dnnl::engine::kind, dnnl::engine>::const_iterator iter = dnnl_engine.find(dnnl::engine::kind::cpu);
+    if (iter != dnnl_engine.end()) {
+      cpu_engine = (dnnl::engine)iter->second;
```
This un-needed C-style cast comes from referencing the setup code currently in use for the other ops.  The cast is un-needed and has been removed.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h:
```diff
@@ -0,0 +1,258 @@
+                        std::vector<std::unordered_map<int, dnnl::memory>>& net_args) override {
+    dnnl::engine cpu_engine;
+    dnnl::engine engine_to_use;
+    std::unordered_map<dnnl::engine::kind, dnnl::engine>::const_iterator iter = dnnl_engine.find(dnnl::engine::kind::cpu);
```
changed.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h:
```diff
@@ -0,0 +1,258 @@
+
+    int input_index = mklnode_ptr_->input_start_index < 0 ? 0 : mklnode_ptr_->input_start_index;
+    if (mklnode_ptr_->parent_nodes.empty()) {
+      //
```
done

georgen117 commented on orttraining/orttraining/test/gradient/gradient_checker.cc:
```diff
@@ -98,7 +98,7 @@ inline std::vector<OrtValue> GradientChecker<X_T, Y_T, JAC_T>::EvaluateFunctionA
     op_session.AddOutput<Y_T>(name.c_str(), y_infos[data_index].shape.GetDims(), (*y_datas)[data_index]);
   }
-
+  //op_session.Run(OpTester::ExpectResult::kExpectSuccess, "", {}, nullptr, execution_providers_);
```
removed.

georgen117 commented on orttraining/orttraining/test/gradient/gradient_ops_test.cc:
```diff
@@ -1,7 +1,7 @@
 // Licensed under the MIT License.
 
-#ifdef NDEBUG  // disable for debug builds because some of these tests are slow
+#ifndef DEBUG  // disable for debug builds because some of these tests are slow
```
I think this was due to a rebase.  I don't recall ever actually modifying this line of code.

reverted change.

georgen117 replied:
Found out one of the other developers working on this with me made the change so he could work on the test in debug mode and it inadvertently made it into the commit. Thanks for the catch.

georgen117 commented on orttraining/orttraining/test/gradient/gradient_ops_test.cc:
```diff
@@ -795,8 +869,142 @@ TEST(GradientCheckerTest, ConvGrad) {
                                           false);
     EXPECT_IS_TINIER_THAN(max_error, error_tolerance);
   }
+    //conv_with_dialation
```
Thanks for the catch.

georgen117 replied:
fixed

georgen117 commented on orttraining/orttraining/test/gradient/gradient_ops_test.cc:
```diff
@@ -709,6 +719,70 @@ TEST(GradientCheckerTest, MaxPoolGrad) {
 }
 
+#ifdef USE_DNNL 
+TEST(GradientCheckerTest, MaxPoolGradDnnl) {
```
Good suggestion. I will see what I can do.

georgen117 replied:
The tests for MaxPoolGrad and ConvGrad have been combined as suggested.

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.cc:
```diff
@@ -129,11 +129,35 @@ void DNNLExecutionProvider::CreateOrUpdateDnnlNode(const Provider_Node* node,
+// operations. Store the index of the node and the the list of input nodes.
+// The input nodes can be used to find the forward pass node.
+// The onnx node index is being used instead of the subgraph index because forwardpass
+// and backward pass nodes are likly to span beyond the subgraph.
```
fixed

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h:
```diff
@@ -0,0 +1,258 @@
+      primitive_dst_mem_ = std::make_shared<dnnl::memory>(dnnl::memory(relu_bwd_pd_.get()->diff_src_desc(), engine_to_use));
+    }
+
+    relugrad_bwd_ = onnxruntime::make_unique<dnnl::eltwise_backward>(
+        dnnl::eltwise_backward(*relu_bwd_pd_));
+
+    net.push_back(*relugrad_bwd_);
```
You have a point. I will have to investigate.

georgen117 replied:
code update based on suggestion

georgen117 commented:
@mrry I want to thank you for all the time you took to review the code. I know how hard it is to look over other developers' code and figure out what is going on.  Your review is really appreciated. I am still working on addressing some of the issues. I hope to have an updated commit soon.

georgen117 commented on orttraining/orttraining/test/gradient/gradient_checker.cc:
```diff
@@ -570,7 +574,11 @@ inline Status GradientChecker<X_T, Y_T, JAC_T>::ComputeGradientError(
+    bool check_not_have_shape_inferencing /* = false*/,
+    std::vector<std::unique_ptr<IExecutionProvider>>* execution_providers /* = nullptr */) {
+
+  execution_providers_ = execution_providers;
```
I originally implemented by adding the `execution_providers` to `ComputeGradientErrorInternal()` and `Co,puteTheoreticalJacobianTranspose()`.  I later changed it.  I have no problem doing as suggested.

georgen117 replied:
Done

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h:
```diff
@@ -0,0 +1,258 @@
+    TensorShape x_shape;
+    if (mklnode_ptr_->parent_nodes.empty()) {
+      // convert the onnx gradient Tensor to dnnl::memory and dnnl::memory::desc
+      const OrtValue* xgrad_input_tensor = ort.KernelContext_GetInput(context, input_index);
```
At the start of the session, the `CreatePrimitives` would be called once for each ReluGrad node in the graph. This will set up the input/output tensors based on the input graph. This only sets up the node it is not executed at this stage. There is an instance of the DnnlReluGrad for each ReluGrad node. The should be no synchronization problems between multiple threads. 

 When the ReluGrad node is executed it will call the `Bind()` function to get the memory from OnnxRuntime then the actual operator is run. (See the `SubgraphPrimitive::Compute(...)` found in dnnl_func_kernel.cc)

mrry replied:
Thanks for bearing with me! This EP is new territory for me, so I took a look at dnnl_func_kernel.cc to understand. It looks like the thread-local maps in `PrimitivePool<T>::GetMap()` are taking care of the thread-safety issue.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h:
```diff
@@ -0,0 +1,258 @@
+
+      xgrad_shape = TensorShape(xgradshape, xgraddim);
+
+      if (xgrad_shape.NumDimensions() == 0) {
+        primitive_created_status_ = Status(common::ONNXRUNTIME,
+                                           common::INVALID_ARGUMENT,
+                                           "Shape of size zero " + xgrad_shape.ToString());
+        return;
+      }
```
Since the input tensor for `ReluGrad` should have the same shape and size as the corresponding `Relu` operator I reused the error logic from the `Relu` operator without giving it much thought.  We will look into this and get back to you.

georgen117 replied:
We tried to produce a test that would use a scalar tensor for Relu or ReluGrad.  We were unable to actually get Relu to run with a scalar tensor.

mrry replied:
Hmm, I'm surprised that Relu wouldn't accept a scalar tensor (it seems to be [implicitly allowed by the ONNX spec](https://github.com/onnx/onnx/blob/ff047ef41d5d91ac1aaa8ebe85ed449c8f0f3604/onnx/defs/math/defs.cc#L484)), although I guess it's not a very common case in NN implementations!

Either way, I still think the test should be on `xgrad_shape.Size()` to guard against returning `nullptr` from `GetTensorData()` below.

georgen117 replied:
After talking it over with developers that have been helping with this code.  I have changed the code to `Size()` as suggested. I have changed it in Relu and ReluGrad since the logic was the same in both locations.

georgen117 commented:
Although I just did a rebase on Tuesday it looks like we have a new conflict. I will rebase again soon.

@mrry there are still a few of your review comments that I have not yet addressed. My co-worker and I are still looking into them.  And we will get back to you.

georgen117 commented:
@mrry I don't know the procedure for items that have been resolved or fixed.  Do I as the contributor click the "Resolve Conversation" button, or do you as the reviewer click on the button?

georgen117 commented on onnxruntime/core/providers/dnnl/dnnl_execution_provider.h:
```diff
@@ -94,6 +114,13 @@ class DNNLExecutionProvider : public Provider_IExecutionProvider {
+  // running in training mode.The backward Kernels need access the forward kernals; typically
+  // to obtain the forward primitive description but it may be need for other items like
+  // accessing workspace memory.
+  std::map<onnxruntime::NodeIndex, std::shared_ptr<ort_dnnl::DnnlKernel>> fwd_kernal_map_;
```
1. verified that a new DnnlExecutionProvider is created per session by creating multiple sessions in the mnist sample.
2. In testing the fwd_kernel_map_ is always accessed by the same thread the entire time the training is running.   The only time it is accessed by a different thread is at the end of the training session when it runs an inference session to check the accuracy of the trained graph. Since they are run sequentially there are never two threads trying to access the map at the same time.

mrry commented on orttraining/orttraining/test/gradient/gradient_op_test_utils.cc:
```diff
@@ -170,6 +215,7 @@ void GradientOpTester::Run(
           }
           if (!valid)
             break;
+          } else {
```
Something seems to have gone wrong with the formatting here (and with the added "`}`" on L233). Is this leftover debugging code that could be reverted?

georgen117 replied:
Not sure what is going on with the empty else statement I will remove it.

mrry commented on orttraining/orttraining/test/gradient/gradient_op_test_utils.cc:
```diff
@@ -184,6 +230,7 @@ void GradientOpTester::Run(
       fetches_ = ExecuteModel<onnxruntime::training::TrainingSession>(*p_model, session_object, expect_result, expected_failure_string, run_options,
                                                                       feeds, output_names, provider_type);
     }
+    }
```
Same comment as above about leftover debugging code.

georgen117 replied:
This is actually a result of bad code formatting the code should be indented differently.

georgen117 replied:
The indentation has been fixed it should be clear that both curly braces are part of the code.

mrry commented on orttraining/orttraining/test/gradient/gradient_op_test_utils.cc:
```diff
@@ -121,6 +121,51 @@ void GradientOpTester::Run(
+          << "Empty execution providers vector.";
+      std::string provider_types;
+
+      for (auto& entry : *execution_providers) {
```
Adding a comment here to make sure we keep this similar to whatever ends up in `OpTester::Run()`.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_func_kernel.cc:
```diff
@@ -70,6 +75,9 @@ class SubgraphPrimitive : public PrimitiveBase {
         std::shared_ptr<DnnlConv<T>> kernel;
         kernel = std::make_shared<DnnlConv<T>>(dnnl_node, params.provider, *params.attributes, os.str());
+#ifdef ENABLE_TRAINING
+        params.provider->fwd_conv_stack.emplace(kernel);
```
@mrry The use of the stack was due to not being able to figure out another way to get the information needed in the convolution gradient operator.

In the Onnx specification, the gradient ops are independent of the corresponding forward op. DNNL on the other hand expects to have a pointer to the forward op.

For ReluGrad and MaxPoolGrad we were able to use the inputs to find the corresponding forward op. For ConvGrad we were not able to do the same.  Ideally, we would be able to use the input weight to the ConvGrad and use that to find the corresponding Conv operator since they both have the same weight as inputs.  I was unable to find a way to actually do that.

> (1.) Is the stack always empty when calling into this method?

I went back and investigated.  If an inference run is done then the ConvGrad will not be run so the stack will not be cleared. We have added code to make sure the stack is cleared every time the graph is created. 

> (1.) Could it be a local variable instead?

The forward Conv and the backward ConvGrad will likely be in separate subgraphs. So it can not be a local variable we need something that can cross between subgraphs.

> (2.) It looks like there's an assumption that ConvGrad nodes will visited in the reverse order from the corresponding Conv nodes. However, is that guaranteed to be true if (e.g.) there is branching in the graph?

Yes, there is an assumption that the ConvGrad nodes will be visited in the reverse order from the corresponding Conv nodes.  We don't actually know if this is guaranteed to be true it has been true in our testing but admittedly we have been limited to only a few models. We are open to suggestions if you have any.

mrry replied:
> Yes, there is an assumption that the ConvGrad nodes will be visited in the reverse order from the corresponding Conv nodes. We don't actually know if this is guaranteed to be true it has been true in our testing but admittedly we have been limited to only a few models.

It's definitely *not* guaranteed to be true. Did you test with an Inception model, or something else that has multiple parallel branches of Conv nodes? 

As a short term patch, you could add validation to ensure that the input and weight inputs are the same for both the Conv and ConvGrad tensors, and if not raise a not-implemented error. You'd still need to try something else to handle the Inception case robustly, but it would avoid producing incorrect output in the meantime.

georgen117 replied:
> As a short term patch, you could add validation to ensure that the input and weight inputs are the same for both the Conv and ConvGrad tensors, and if not raise a not-implemented error. You'd still need to try something else to handle the Inception case robustly, but it would avoid producing incorrect output in the meantime.

Thanks for the suggestion I will implement it.  I will try and have an update by end of the day Monday. 

Do you have any suggestions on a more robust way we could get a pointer to the forward operator?

georgen117 replied:
I have added a check to the ConvGrad code that checks the tensors from the forward pass against the tensors in the backward pass. Feel free to check the code out dnnl_convgrad.h lines 194-215.

Sorry I didn't get this in on Monday as expected.

georgen117 commented on orttraining/orttraining/test/gradient/gradient_ops_test.cc:
```diff
@@ -795,8 +869,142 @@ TEST(GradientCheckerTest, ConvGrad) {
+}
+
+#ifdef USE_DNNL
+TEST(GradientCheckerTest, ConvGradDnnl) {
```
The ConvGradDnnl and ConvGrad tests have been combined.

georgen117 commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_relugrad.h:
```diff
@@ -0,0 +1,258 @@
+                                           common::INVALID_ARGUMENT,
+                                           "Shape of size zero " + x_shape.ToString());
+        return;
+      }
```
Change made as commented above.

georgen117 commented:
@mrry I have tried to address all of the review requests. Please let me know if there are any additional questions or requests that are needed.

georgen117 commented:
@mrry it has been 7 days since I asked if there are any additional comments or feedback.  I was wondering if there were any additional feedback to get this pull request to move forward.

mrry commented on onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h:
```diff
@@ -191,6 +191,30 @@ class DnnlConvGrad : public DnnlKernel {
+    dnnl::memory::desc conv_fwd_src_desc_ = conv_fwd_->GetPrimitiveDesc()->src_desc();
+    dnnl::memory::desc conv_fwd_weights_desc_ = conv_fwd_->GetPrimitiveDesc()->weights_desc();
+
+    if (conv_fwd_dst_desc_.dims() != dy_dims ||
```
Is it possible to compare based on the *identity* of the input/weight tensors, rather than just their sizes? I'm not sure if this is feasible, and if not then adding a TODO would suffice. 

(If we happened to have two parallel convolutions that took the same sized tensors, we might inadvertently run an incorrect graph... it should be "safe" in the sense that all the buffers are the right size, but the output would be bogus.)

georgen117 replied:
I don't think I have access to the input/weight tensors _identity_. I will take a look. If I do have access to the _identity_ I may be able to remove the stack.

georgen117 replied:
@mrry do you know if the name of the input weight is unique to each input weight?  If so I could use the name of the weight to identify the tensors.

georgen117 replied:
@mrry in response to you asking me to see if it was possible to compare based on the _identity_ of the input/weight tensors I did some investigating. 

Originally I was unable to use the weight because the weight name was not filled in for ConvGrad node. I found the location in the code that the weight name should have been assigned for the ConvGrad and with that change I was able to remove the stack that was being used and replace it with a map.

This basically matches how I originally wanted to implement the code and your simple question about the _identity_ of the input/weight tensors helped me find the correct location in the code to make the change.

Thanks!

mrry replied:
That's wonderful news! Thanks for persevering with my awkward questions :).

georgen117 replied:
Your questions and requests did not but improve the code. Thanks for taking the time to review this huge commit.

mrry reviewed (changes requested):
Thanks George! Just one last question about the error checking and then this should be good to merge.

mrry reviewed (approved):
Thanks George! Looks good to me.

georgen117 commented:
> Thanks George! Looks good to me.

Thanks, Derek, what are my next steps? Am I just waiting on the CI pipeline or do we need to contact someone else?

mrry commented:
As far as I know, all we need to do now is wait for the CI pipelines to come back successful and we can merge.

@jywu-msft I notice you manually ran a DNNL pipeline back at the start of the PR... is that something we need to trigger manually?

mrry commented:
/azp run orttraining-linux-ci-pipeline

mrry commented:
/azp run Windows GPU CI Pipeline, Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline

mrry commented:
/azp run Windows GPU TensorRT CI Pipeline, centos7_cpu, orttraining-linux-gpu-ci-pipeline, Android CI Pipeline, Linux DNNL CI Pipeline, iOS CI Pipeline, orttraining-win-ci-pipeline, orttraining-win-gpu-ci-pipeline, orttraining-mac-ci-pipeline

mrry commented:
/azp run Linux DNNL CI Pipeline

mrry commented:
/azp orttraining-distributed, Linux Nuphar CI Pipeline

mrry commented:
/azp run orttraining-distributed, Linux Nuphar CI Pipeline

mrry commented:
/azp run Android CI Pipeline

georgen117 commented:
@mrry thanks for kicking off all the CI pipelines.  For future reference is that something I could have done or does it need to be someone with Member permissions?

Let me know if there is anything else that I can do to help get this merged.

mrry commented:
No problem, George! I think you need to be a member of the onnxruntime team to kick them off, and there are some circumstances in which they are kicked off automatically, but I had to do it manually this time :).

This is ready to merge now, so I'll just click on the button. Have a great weekend!
